### PR TITLE
Refactor extension documentation

### PR DESCRIPTION
**Pull request by kolchfa-aws**
### Description
Refactor extension documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by dbwiddis — DISMISSED** [review]
Thanks for this contribution! Great readability improvements!

File: PLUGIN_MIGRATION.md [hunk 1]
```diff
@@ -2,49 +2,51 @@
...
-   - The `createComponents()` method no longer takes parameters. The parameters formerly sent in a constructor may be accessed using Guice `@Inject` annotation.
-   - Extension developers who need to use a Rest Client should initialize it using `SDKClient` and return the appropriate client as an object in `createComponents()` to make it available for extension actions.
-   - Annotations of `@Inject` in actions bound using `getActions()` should change the import from the OpenSearch internal package to `com.google.inject.Inject`
+## Implement extension interfaces and extension points
 
-## Use Wrapper Classes
+ - Change the implementing class `FooPlugin` to `FooExtension` and either implement the _`Extension`_ interface or extend `BaseExtension`.
+ - Implement other corresponding extension interfaces, for example `ActionPlugin` interface becomes `ActionExtension`.
+ - Change extension point implementation, if necessary, to conform to new types. For example, `RestHandler` classes become `ExtensionRestHandler` classes. Consider extending base implementations such as `BaseExtensionRestHandler` to provide additional convenience methods.
```
Comment (natebower) [inline PLUGIN_MIGRATION.md:18]:
```suggestion
 - Change extension point implementation, if necessary, to conform to new types. For example, `RestHandler` classes become `ExtensionRestHandler` classes. Consider extending base implementations, such as `BaseExtensionRestHandler`, to provide additional convenience methods.
```

File: PLUGIN_MIGRATION.md [hunk 2]
```diff
@@ -2,49 +2,51 @@
...
+ - The `createComponents()` method no longer takes parameters. The parameters formerly sent in a constructor may be accessed using Guice `@Inject` annotation.
+ - Extension developers who need to use a REST client should initialize it using `SDKClient` and return the appropriate client as an object in `createComponents()` to make it available for extension actions.
+ - Annotations of `@Inject` in actions bound using `getActions()` should change the import from the OpenSearch internal package to `com.google.inject.Inject`.
 
-### Replace ClusterService with SDKClusterService
+## Use wrapper classes
```
Comment (natebower) [inline PLUGIN_MIGRATION.md:23]: Confirm that these headings will not end up stacked.

File: PLUGIN_MIGRATION.md [hunk 3]
```diff
@@ -2,49 +2,51 @@
...
  - Calls to `clusterService.state()` do not require changes.
 
-### Replace Client with SDK Client, either OpenSearchClient (JavaClient) or SDKRestClient
+### Replace `Client` with `SDKClient`, either `OpenSearchClient` (`JavaClient`) or `SDKRestClient`
 
 The `SDKClient` provides two (eventually three) client options.
 
-The [Java Client for OpenSearch](https://github.com/opensearch-project/opensearch-java) (`OpenSearchClient`) will be supported with both synchronous and asynchronous clients, and is actively developed along with other language clients and should be used whenever possible. These clients do have significant implementation differences compared to the existing `Client` interface implemented by plugins.
+The [Java client for OpenSearch](https://github.com/opensearch-project/opensearch-java) (`OpenSearchClient`) will be supported with both synchronous and asynchronous clients, and is actively developed along with other language clients and should be used whenever possible. These clients do have significant implementation differences compared to the existing `Client` interface implemented by plugins.
```
Comment (natebower) [inline PLUGIN_MIGRATION.md:34]:
```suggestion
The [Java client for OpenSearch](https://github.com/opensearch-project/opensearch-java) (`OpenSearchClient`) will be supported with both synchronous and asynchronous clients, is actively developed along with other language clients, and should be used whenever possible. These clients do have significant implementation differences compared to the existing `Client` interface implemented by plugins.
```

File: README.md [hunk 4]
```diff
@@ -3,29 +3,39 @@
...
+  * [Creating an extension](#creating-an-extension)
+  * [Developer Guide](#developer-guide)
+  * [Plugin migration](#plugin-migration)
+  * [Contributing](#contributing)
+  * [Maintainers](#maintainers)
+  * [Code of Conduct](#code-of-conduct)
 
 ## Introduction
-Opensearch plugins have allowed the extension and ehancements of various core features however, current plugin architecture carries the risk of fatally impacting clusters should they fail. In order to ensure that plugins may run safely without impacting the system, our goal is to effectively isolate plugin interactions with OpenSearch by modularizing the [extension points](https://opensearch.org/blog/technical-post/2021/12/plugins-intro/) to which they hook onto.
+With OpenSearch plugins, you can extend and enhance various core features. However, the current plugin architecture may fatally impact clusters in case of failure. To ensure that plugins run safely without impacting the system, our goal is to isolate plugin interactions with OpenSearch. The OpenSearch SDK for Java modularizes the [extension points](https://opensearch.org/blog/technical-post/2021/12/plugins-intro/) onto which plugins hook.
```
Comment (natebower) [inline README.md:18]: Third sentence: Should "with OpenSearch" be "within OpenSearch"?

**Review by natebower — COMMENTED** [review]
@kolchfa-aws Please see my comments and changes and let me know if you have any questions. Thanks!

**Review by natebower — COMMENTED** [review]
@kolchfa-aws Please see my comments and changes and let me know if you have any questions. Thanks!

**Comment by kolchfa-aws** [comment]
Ready for review/merge. Thanks!

**Review by ryanbogan — DISMISSED** [review]

**Comment by kolchfa-aws** [comment]
@dbwiddis Two questions about the `DESIGN` document's "Responding to REST requests" section:
- In step (19), there is a reference to `RestExecuteOnExtensionRequest`. Has this class been renamed?
- In step (20), there is a reference to `pathMap` which I believe is also outdated. Could you provide a suggestion for rewording please?

**Comment by dbwiddis** [comment]
> * In step (19), there is a reference to `RestExecuteOnExtensionRequest`. Has this class been renamed?

The method is `handleRestExecuteOnExtensionRequest()`:
https://github.com/opensearch-project/opensearch-sdk-java/blob/04d9eab8ca08acca572f258053e05e18809adf4b/src/main/java/org/opensearch/sdk/handlers/ExtensionsRestRequestHandler.java#L57

So in the context of step 19 that should be `ExtensionRestRequest`.

> * In step (20), there is a reference to `pathMap` which I believe is also outdated. Could you provide a suggestion for rewording please?

This is now a variable `pathTrie` although a more generic term "path registry" may be appropriate to use here.  Also where it says "The `Extension` matches" should probably be "The `ExtensionRestPathRegistry` matches".

**Review by ryanbogan — APPROVED** [review]

**Comment by owaiskazi19** [comment]
@kolchfa-aws can you rebase with the latest `main`, resolve the conflicts and let's get this PR in?

**Comment by kolchfa-aws** [comment]
@owaiskazi19 Done. Should be ready to merge. Thank you.

**Comment by kolchfa-aws** [comment]
Done.

**Review by dbwiddis — APPROVED** [review]

**Review by owaiskazi19 — APPROVED** [review]

**Comment by kolchfa-aws** [comment]
Done #816

**Pull request merged**